### PR TITLE
Clean up types

### PR DESCRIPTION
**Pull request by wch**
This PR cleans up types for `Tag` functions. This will reduce ambiguities in functions that call Tag functions, and simplifies the API a bit.

The biggest change is that, instead of a single `TagChildArg` type that encompasses both children and attribute dictionaries, it splits it into two types:
* `TagNode` for actual children, like strings, numbers, `Tag` objects, and lists of those objects.
* `TagAttrs`, for dictionaries of attributes.

The `Tag` functions have a signature like:

```py
def div(*args: TagChild | TagAttrs, **kwargs: TagAttrValue) -> Tag:
    ...
```

Note that `TagAttrValue` and `TagAttrs` are defined like this:

```py
TagAttrValue = Union[str, float, bool, None]

TagAttrs = Dict[str, TagAttrValue]
```

*****

With this change, functions which wrap Tag functions can be more specific about what they accept. For example, `shiny.ui.input_text` previously looked like this:

```py
def input_text(id: str, label: TagChildArg):
    ...
```

Previously, when `TagChildArg` encompassed both children and attributes, then this allowed `label` to be an attribute dictionary (as in `label={"class":"foo"}`), which was incorrect. Now, since `TagChild` only encompasses children (and not attributes), the function signature accurately reflects what is really allowed.

If a wrapper function really wants to accept both children and attributes, it needs a signature that explicitly accepts both types:

```py
def div_wrapper(*args: TagChild | TagAttrs):
    return div({"class":"wrapper"}, *args)
```

*****

This PR also does the following:

* Drops the `children` parameter from the `Tag` class and tag functions, because it is not used. We previously thought it would be useful when you wanted to put attributes before children, like this:
    ```py
    div(class_="a", children=["b"])
    ```
    But it turns out that it's actually simpler to use a dictionary, like one of the following:
    ```py
    div(dict(class_="a"), "b")
    div({"class": "a"}, "b")
    ```

### CODE REVIEW

File: htmltools/_jsx.py [hunk 1]
```diff
@@ -328,7 +326,7 @@ def jsx_tag_create(
     def create_tag(
         *args: TagChildArg,
         children: Optional[list[TagChildArg]] = None,
```
Comment (wch) [inline htmltools/_jsx.py:328]: Remove `children`

File: htmltools/_core.py [hunk 2]
```diff
@@ -89,23 +90,38 @@ class MetadataNode:
...
-    Sequence["TagChildArg"],
+    Sequence["TagChild"],
 ]
 
 
+# These two types existed in htmltools 0.14.0 and earlier. They are here so that
+# existing versions of Shiny will be able to load, but users of those existing packages
+# will see type errors, which should encourage them to upgrade Shiny.
+TagChildArg = Never
+TagAttrArg = Never
```
Comment (cpsievert) [inline htmltools/_core.py:122]: Just curious, would it be possible to throw a warning if they do actually get imported?

File: htmltools/_core.py [hunk 3]
```diff
@@ -392,14 +409,19 @@ class Tag:
     """
     The HTML tag class.
 
+    A Tag object consists of a name, attributes, and children. The name is a string, the
+    attributes are held in a TagAttrDict object, and the children are held in a TagList
+    object.
+
+    This class usually should not be instantiated directly. Instead, use the tag wrapper
+    functions in ``htmltools.tags``, like ``div()`` or ``a()``.
```
Comment (cpsievert) [inline htmltools/_core.py:417]: I think technically this should use RST syntax to link to the class/func/type, like this ...

https://github.com/rstudio/py-shiny/blob/d5855b0189d1d2385bf622f73edd3bb68bf8f125/shiny/ui/_input_date.py#L45

... but I doubt we're doing that a good job of that in this repo, so I feel free to ignore :)

**Review by cpsievert — APPROVED** [review]
Nice, thank you!

File: htmltools/_core.py (hunk 2 above)
Comment (wch) [reply htmltools/_core.py:122]: Not that I know of. I saw this on SO, but it throws warnings a bit too easily. (https://stackoverflow.com/a/60486106/412655)

```py
class TagChildArg:
   raise DeprecationWarning("text here")
```

The problem is that it raises the warning even if you just do `from htmltools import div`.

File: htmltools/_core.py (hunk 3 above)
Comment (wch) [reply htmltools/_core.py:417]: Good catch. We're going to do a pass over this repo fixing docs once we get quartodoc working properly, so I'll just save it for then.

**Pull request merged**